File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/volume/preparation/MaxFileVolumePreparer.java
package com.lasthopesoftware.bluewater.client.playback.file.volume.preparation;

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PlayableFilePreparationSource;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PreparedPlayableFile;
import com.lasthopesoftware.bluewater.client.playback.file.volume.ProvideMaxFileVolume;
import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.handoff.promises.response.VoidResponse;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class MaxFileVolumePreparer implements PlayableFilePreparationSource {

	private static final Logger logger = LoggerFactory.getLogger(MaxFileVolumePreparer.class);

	private final PlayableFilePreparationSource playableFilePreparationSource;
	private final ProvideMaxFileVolume provideMaxFileVolume;

	public MaxFileVolumePreparer(PlayableFilePreparationSource playableFilePreparationSource, ProvideMaxFileVolume provideMaxFileVolume) {
		this.playableFilePreparationSource = playableFilePreparationSource;
		this.provideMaxFileVolume = provideMaxFileVolume;
	}

	@Override
	public Promise<PreparedPlayableFile> promisePreparedPlaybackFile(ServiceFile serviceFile, long preparedAt) {
		final Promise<Float> promisedMaxFileVolume = provideMaxFileVolume.promiseMaxFileVolume(serviceFile);

		return playableFilePreparationSource
			.promisePreparedPlaybackFile(serviceFile, preparedAt)
			.then(ppf -> {
				final MaxFileVolumeManager maxFileVolumeManager = new MaxFileVolumeManager(ppf.getPlayableFileVolumeManager());

				promisedMaxFileVolume
					.then(new VoidResponse<>(maxFileVolumeManager::setMaxFileVolume))
					.excuse(new VoidResponse<>(err -> logger.warn("There was an error getting the max file volume for file " + serviceFile, err)));

				return new PreparedPlayableFile(
					ppf.getPlaybackHandler(),
					maxFileVolumeManager,
					ppf.getBufferingPlaybackFile());
			});
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/volume/preparation/MaxFileVolumePreparer.kt
package com.lasthopesoftware.bluewater.client.playback.file.volume.preparation

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PlayableFilePreparationSource
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PreparedPlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.volume.ProvideMaxFileVolume
import com.namehillsoftware.handoff.promises.Promise
import org.joda.time.Duration

class MaxFileVolumePreparer(
	private val inner: PlayableFilePreparationSource,
	private val provideMaxFileVolume: ProvideMaxFileVolume
) : PlayableFilePreparationSource {
	override fun promisePreparedPlaybackFile(libraryId: LibraryId, serviceFile: ServiceFile, preparedAt: Duration): Promise<PreparedPlayableFile?> = Promise.Proxy { cp ->
		val promisedPreparedFile = inner.promisePreparedPlaybackFile(libraryId, serviceFile, preparedAt).also(cp::doCancel)
		provideMaxFileVolume
			.promiseMaxFileVolume(libraryId, serviceFile)
			.also(cp::doCancel)
			.eventually { maxFileVolume ->
				promisedPreparedFile
					.then { ppf ->
						ppf ?: return@then null

						val maxFileVolumeManager = MaxFileVolumeManager(ppf.playableFileVolumeManager)
						maxFileVolumeManager.setMaxFileVolume(maxFileVolume)

						PreparedPlayableFile(
							ppf.playbackHandler,
							maxFileVolumeManager,
							ppf.bufferingPlaybackFile
						)
					}
			}
	}
}
